tools/run_block.py

```diff
@@ -0,0 +1,187 @@
+        "conditions": a list of condition expressions, as in [condition_opcodes.py](../chia/types/condition_opcodes.py)
+        "puzzle_hash": the sha256 of the CLVM bytecode that controls spending this coin
+
+Condition Opcodes, such as AGG_SIG_ME, or CREATE_COIN are created by running the "puzzle", i.e. the CLVM bytecode
+associated with the coin being spent. Condition Opcodes are verified by every client on the network for every spend,
+and in this way they control whether a spend is valid or not.
+
+"""
+import json
+from dataclasses import dataclass
+from typing import List, TextIO, Tuple
+
+import click
+from chia.consensus.constants import ConsensusConstants
+from chia.consensus.default_constants import DEFAULT_CONSTANTS
+from chia.full_node.mempool_check_conditions import get_name_puzzle_conditions, get_puzzle_and_solution_for_coin
+from chia.types.blockchain_format.program import SerializedProgram
+from chia.types.condition_opcodes import ConditionOpcode
+from chia.types.condition_with_args import ConditionWithArgs
+from chia.types.full_block import FullBlock
+from chia.types.generator_types import BlockGenerator, GeneratorArg
+from chia.types.name_puzzle_condition import NPC
+from chia.util.config import load_config
+from chia.util.default_root import DEFAULT_ROOT_PATH
+from chia.util.errors import ConsensusError, Err
+from chia.util.ints import uint32
+from chia.wallet.cat_wallet.cat_utils import match_cat_puzzle
+
+
+@dataclass
+class CAT:
+    tail_hash: str
+    memo: str
+    npc: NPC
+
+    def cat_to_dict(self):
+        return {"tail_hash": self.tail_hash, "memo": self.memo, "npc": npc_to_dict(self.npc)}
+
+
+def condition_with_args_to_dict(condition_with_args: ConditionWithArgs):
+    return {
+        "condition_opcode": condition_with_args.opcode.name,
+        "arguments": [arg.hex() for arg in condition_with_args.vars],
+    }
+
+
+def condition_list_to_dict(condition_list: Tuple[ConditionOpcode, List[ConditionWithArgs]]):
+    assert all([condition_list[0] == cwa.opcode for cwa in condition_list[1]])
+    return [condition_with_args_to_dict(cwa) for cwa in condition_list[1]]
+
+
+def npc_to_dict(npc: NPC):
+    return {
+        "coin_name": npc.coin_name.hex(),
+        "conditions": [{"condition_type": c[0].name, "conditions": condition_list_to_dict(c)} for c in npc.conditions],
+        "puzzle_hash": npc.puzzle_hash.hex(),
+    }
+
+
+def run_generator(block_generator: BlockGenerator, constants: ConsensusConstants) -> List[CAT]:
+    npc_result = get_name_puzzle_conditions(
+        block_generator,
+        constants.MAX_BLOCK_COST_CLVM,  # min(self.constants.MAX_BLOCK_COST_CLVM, block.transactions_info.cost),
+        cost_per_byte=constants.COST_PER_BYTE,
+        mempool_mode=False,
+    )
+    if npc_result.error is not None:
+        raise ConsensusError(Err(npc_result.error))
+
+    cat_list: List[CAT] = []
+    for npc in npc_result.npc_list:
+        _, puzzle, solution = get_puzzle_and_solution_for_coin(
```

Comment:
unrelated to this patch, but `get_puzzle_and_solution_for_coin()` seems to run the whole block program again, which isn't ideal. But I suppose that's mostly an implementation detail for now.

Reply:
Yes, I think this tool ends up running twice.